src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionManager.cs

```diff
@@ -0,0 +1,159 @@
+                _timeout = timeout;
+                _remotableDataRpc = remotableDataRpc;
+
+                _maxPoolConnections = maxPoolConnection;
+
+                // we have 4 services. so start from 4. later if we add more services, it will still work.
+                _pools = new ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>>(concurrencyLevel: 4, capacity: 4);
+
+                _usePool = usePool;
+                _shutdownLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+            }
+
+            public Task<Connection> TryCreateConnectionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
+            {
+                // pool is turned off either by option or pool has shutdown.
+                if (!_usePool)
+                {
+                    // RemoteHostClient is allowed to be restarted by IRemoteHostClientService.RequestNewRemoteHostAsync
+                    // when that happen, existing RemoteHostClient doesn't suddenly go away and start to throw exception. it handles
+                    // shutdown gracefully. it will keep serving existing requests or even new requests if someone is holding old remoteHostClient.
+                    // all connections made to old RemoteHost will eventually go away and that's when RemoteHost is actually removed.
+                    // simplyput, RequestNewRemoteHostAsync is not intrusive to running features. all existing one will keep do what is doing
+                    // with old RemoteHost and only new request will go to new remoteHost.
+                    return TryCreateNewConnectionAsync(serviceName, callbackTarget, cancellationToken);
+                }
+
+                // when callbackTarget is given, we can't share/pool connection since callbackTarget attaches a state to connection.
+                // so connection is only valid for that specific callbackTarget. it is up to the caller to keep connection open
+                // if he wants to reuse same connection
+                if (callbackTarget != null)
+                {
+                    return TryCreateNewConnectionAsync(serviceName, callbackTarget, cancellationToken);
+                }
+
+                return TryGetConnectionFromPoolAsync(serviceName, cancellationToken);
+            }
+
+            private async Task<Connection> TryGetConnectionFromPoolAsync(string serviceName, CancellationToken cancellationToken)
+            {
+                var queue = _pools.GetOrAdd(serviceName, _ => new ConcurrentQueue<JsonRpcConnection>());
+                if (queue.TryDequeue(out var connection))
+                {
+                    return new PooledConnection(this, serviceName, connection);
+                }
+
+                return new PooledConnection(this, serviceName, (JsonRpcConnection)await TryCreateNewConnectionAsync(serviceName, callbackTarget: null, cancellationToken).ConfigureAwait(false));
```

Comment:
can't use Task<JsonRpcConnection> since generic on type doesnt support covariant. and I want PooledConnection to have strongly typed one since it only allow JsonRpcConnection type. and I am not sure why casting is bad here.

Reply:
❓  `TryCreateNewConnectionAsync` is a private method. Why can the return type not change?

Reply:
Task<JsonRpc...> can be assigned from Task<Connection>

Reply:
well. my point is I dont see any issue of casting. not sure why you want to remove it.

Reply:
Coupling of this form causes a small maintainability hit (technical debt). Sometimes it's unavoidable, but it's much easier to head these off where they start than try to go back and fix the design later.

Reply:
if there is simple way to remove cast then sure, otherwise, I would choose straightforward cast over other more complex form.

underlying issue is simple, if Task < .T > supported covariant and I can do Task< BaseType > = Task< DerivedType > as I can do with interface, I can change TryCreateNewConnectionAsync to return Task< JsonRpc... > and leave TryCreateConnectionAsync to return Task < BaseType >, but unfortunately that is not possible, so my option is either using await in TryCreateConnectionAsync even though it is not necessary just to remove cast, or using explicit cast like I did. 

let's see whether this actually become an issue where it becomes hard to find bug.